zpages/rpcz.go

```diff
@@ -205,113 +189,112 @@ func (s snapExporter) ExportView(vd *view.Data) {
+			continue
 		}
 
-		// Update field of s corresponding to the view.
+		var hour, minute *windowStat
 		switch vd.View {
-		case ocgrpc.ClientCompletedRPCsView:
-			if _, ok := haveResetErrors[method]; ok {
-				haveResetErrors[method] = struct{}{}
-				s.ErrorsTotal = 0
-			}
-			for _, tag := range row.Tags {
-				if tag.Key == ocgrpc.KeyClientStatus && tag.Value != "OK" {
-					s.ErrorsTotal += int(count)
-				}
-			}
-
 		case ocgrpc.ClientRoundtripLatencyView:
-			s.AvgLatencyTotal = convertTime(sum / count)
-
+			hour = &s.LatencyHour
+			minute = &s.LatencyMinute
 		case ocgrpc.ClientSentBytesPerRPCView:
-			s.OutputRateTotal = computeRate(0, sum)
-
+			hour = &s.OutputHour
+			minute = &s.OutputMinute
 		case ocgrpc.ClientReceivedBytesPerRPCView:
-			s.InputRateTotal = computeRate(0, sum)
-
-		case ocgrpc.ClientSentMessagesPerRPCView:
-			s.CountTotal = int(count)
-			s.RPCRateTotal = computeRate(0, count)
-
-		case ocgrpc.ClientReceivedMessagesPerRPCView:
-			// currently unused
-
-		case ocgrpc.ServerCompletedRPCsView:
-			if _, ok := haveResetErrors[method]; ok {
-				haveResetErrors[method] = struct{}{}
-				s.ErrorsTotal = 0
-			}
-			for _, tag := range row.Tags {
-				if tag.Key == ocgrpc.KeyServerStatus && tag.Value != "OK" {
-					s.ErrorsTotal += int(count)
-				}
-			}
-
+			hour = &s.InputHour
+			minute = &s.InputMinute
 		case ocgrpc.ServerLatencyView:
-			s.AvgLatencyTotal = convertTime(sum / count)
-
+			hour = &s.LatencyHour
+			minute = &s.LatencyMinute
 		case ocgrpc.ServerSentBytesPerRPCView:
-			s.OutputRateTotal = computeRate(0, sum)
+			hour = &s.OutputHour
+			minute = &s.OutputMinute
+		case ocgrpc.ServerReceivedBytesPerRPCView:
+			hour = &s.InputHour
+			minute = &s.InputMinute
+		}
+		if hour == nil || minute == nil {
+			continue
+		}
+		hour.update(vd.End, dist)
+		minute.update(vd.End, dist)
+	}
 
-		case ocgrpc.ServerReceivedMessagesPerRPCView:
-			s.CountTotal = int(count)
-			s.RPCRateTotal = computeRate(0, count)
+}
+
+func getStatus(row *view.Row) string {
+	for _, tag := range row.Tags {
+		if tag.Key == ocgrpc.KeyServerStatus || tag.Key == ocgrpc.KeyClientStatus {
+			return tag.Value
+		}
+	}
+	return ""
+}
 
-		case ocgrpc.ServerSentMessagesPerRPCView:
-			// currently unused
+func getMethod(row *view.Row) string {
+	for _, tag := range row.Tags {
+		if tag.Key == ocgrpc.KeyClientMethod || tag.Key == ocgrpc.KeyServerMethod {
+			return tag.Value
 		}
 	}
+	return ""
+}
+
+func snapshot(method string, received bool) *statSnapshot {
+	key := methodKey{method: method, received: received}
+	if s := snaps[key]; s != nil {
+		return s
+	}
+	s := &statSnapshot{Method: method, Received: received}
+	for _, hourWindow := range []*windowStat{&s.OutputHour, &s.LatencyHour, &s.ErrorsHour, &s.InputHour} {
+		hourWindow.init(time.Hour)
+	}
+	for _, minuteWindow := range []*windowStat{&s.OutputMinute, &s.LatencyMinute, &s.ErrorsMinute, &s.InputMinute} {
+		minuteWindow.init(time.Minute)
+	}
+	snaps[key] = s
```

Comment:
Should we be protecting this with a mutex?

Reply:
It already is higher up in the call stack in ExportView